File: add_flight.py
import yaml
import re
from datetime import datetime

def is_valid_iata(code):
    return re.fullmatch(r'[A-Za-z0-9]*[A-Za-z][A-Za-z0-9]*\d{1,4}', code) is not None
def is_valid_date(date):
    # Check if the date matches the format YYYY-MM-DD
    try:
        datetime.strptime(date, '%Y-%m-%d')
        return True
    except ValueError:
        return False
def add_flights():
    with open('config.yaml', 'r') as file:
        data = yaml.safe_load(file)

    user_name = input("Enter your name: ")
    user_id = input("Enter your user ID: ")

    print("Select an event:")
    for i, event in enumerate(data['events']):
        print(f"{i+1} for {event['name']}")
    while True:
        try:
            event_index = int(input("Enter your choice: ")) - 1
            if 0 <= event_index < len(data['events']):
                break
            else:
                print("Invalid choice. Please try again.")
        except ValueError:
            print("Invalid input. Please enter a number.")

    user = {'id': user_id, 'name': user_name, 'flights': []}
    data['events'][event_index]['users'].append(user)

    while True:
        while True:
            flight_code = input("Enter your flight code (or 'q' to quit): ")
            if flight_code.lower() == 'q' or is_valid_iata(flight_code):
                break
            else:
                print("Invalid flight code. Please try again.")
        if flight_code.lower() == 'q':
            break
        while True:
            flight_date = input("Enter your flight date (YYYY-MM-DD): ")
            if is_valid_date(flight_date):
                break
            else:
                print("Invalid date format. Please try again.")

        user['flights'].append({
            'code': flight_code,
            'date': flight_date
        })

    with open('config.yaml', 'w') as file:
        yaml.safe_dump(data, file)

add_flights()

File: interactions/addFlight.js
const yaml = require('js-yaml');
const fs = require('fs');
const config = yaml.load(fs.readFileSync('./config.yaml', 'utf8'));
const { PrismaClient } = require("@prisma/client");
const { App } = require('@slack/bolt');
const prisma = new PrismaClient();
const utils = require("../utils")
/**
 * @param {App} app
 */
module.exports = (app, pull) => {
    app.view('add_flight', async ({ view, ack, body }) => {
        await ack()
        const submittedValues = view.state.values
        var json = JSON.parse(view.private_metadata)
        //console.log(submittedValues)
        let flightNumber, flightDate, event
        for (let key in submittedValues) {
            if (submittedValues[key]['flight_number']) flightNumber = submittedValues[key]['flight_number'].value.toUpperCase()
            if (submittedValues[key]['flight_date']) flightDate = new Date(submittedValues[key]['flight_date'].selected_date)
            if (submittedValues[key]['select_event']) event = submittedValues[key]['select_event'].selected_option.value

        }
        if (!flightNumber || !flightDate || !event) return await app.client.chat.postEphemeral({
            user: body.user.id,
            channel: json.channel_id,
            text: "All fields are required"
        })
        const existingFlight = await prisma.flight.findFirst({
            where: {
                slackId: body.user.id,
                flightNumber
            }
        })
        if (existingFlight) return await app.client.chat.postEphemeral({
            user: body.user.id,
            channel: json.channel_id,
            text: "You already have this as an existing flight."
        })

        // Note from David: Flights can be added more than two days in advanced.
        // Slight issue: The aeroapi only supports up to 2 days in advanced.
        // However, codes are usually still valid even weeks before hand, so we check with today's date

        const flight = await utils.checkFlight(flightNumber, new Date())
        if (!flight) return await app.client.chat.postEphemeral({
            user: body.user.id,
            channel: json.channel_id,
            text: "The flight code given is not valid."
        })
        await prisma.flight.create({
            data: {
                slackId: body.user.id,
                flightNumber,
                flightDate: flightDate,
                event
            }
        })
        await pull()

        return await app.client.chat.postEphemeral({
            user: body.user.id,
            channel: json.channel_id,
            text: "Added flight successfully."
        })
    })
    app.command('/addflight', async ({ ack, body, client, logger }) => {
        await ack();
        await client.views.open({
            trigger_id: body.trigger_id,
            view: {
                private_metadata: JSON.stringify({ channel_id: body.channel_id }),
                callback_id: 'add_flight',
                "title": {
                    "type": "plain_text",
                    "text": "Add your flight",
                    "emoji": true
                },
                "submit": {
                    "type": "plain_text",
                    "text": "Submit",
                    "emoji": true
                },
                "type": "modal",
                "close": {
                    "type": "plain_text",
                    "text": "Cancel",
                    "emoji": true
                },
                "blocks": [
                    {
                        "type": "section",
                        "text": {
                            "type": "mrkdwn",
                            "text": "Add your own flight to the tracking board."
                        }
                    },
                    {
                        "type": "input",
                        "element": {
                            "type": "plain_text_input",
                            "action_id": "flight_number"
                        },
                        "label": {
                            "type": "plain_text",
                            "text": "Flight Number",
                            "emoji": true,
                        },
                        optional: false
                    },
                    {
                        "type": "section",
                        "text": {
                            "type": "mrkdwn",
                            "text": "Select the date of your flight"
                        },
                        "accessory": {
                            "type": "datepicker",
                            "placeholder": {
                                "type": "plain_text",
                                "text": "Select a date",
                                "emoji": true,
                            },
                            "action_id": "flight_date",
                        },

                    },
                    {
                        "type": "section",
                        "text": {
                            "type": "mrkdwn",
                            "text": "Select the event"
                        },
                        "accessory": {
                            "type": "static_select",
                            "placeholder": {
                                "type": "plain_text",
                                "text": "Select an item",
                                "emoji": true,
                            },
                            "options": config.events.map(event => {
                                return {
                                    "text": {
                                        "type": "plain_text",
                                        "text": event.name,
                                        "emoji": true
                                    },
                                    "value": event.name
                                }
                            }),
                            "action_id": "select_event",
                        }
                    }
                ]
            }
        })
    })
}